rfcs/037-index-query-composition.md

```diff
@@ -0,0 +1,115 @@
+
+- What tables are most commonly joined to this table and how?
+- Are there any filters (where clauses) that are always applied when accessing this table?
+- What real-world queries are being used on this table?
+- What queries are relevant now, given the recent queries that have been executed?
+
+This information helps users to discover _how to use_ their data.
+
+## Guide-level Explanation (aka Product Details)
+
+Four new asset types are added to Amundsen as depicted in the following diagram:
+
+![Query Assets](../assets/037/query-assets.png)
+
+1. **Query**: these contain the full SQL for a query as well as relationships to Tables, Users, Query Executions, Joins and Wheres
+2. **Query Execution**: this represents the total number of times a query was executed within a time window. By having query executions separated from queries Amundsen can keep one copy of the query in the database and sum the total number of executions at any point in time to get the total execution count. Query Executions can be added or dropped to add or remove relevance to a Query, Join or Where
+3. **Join**: this is a component of a query, a single join clause (e.g. `table_1 a join table_2 b on a.field = b.field`). Joins are connected to the columns they are joined on as well as to queries. By connecting to queries, Amundsen can rank order the most frequent joins by accessing the query executions.
+4. **Where**: this is a component of a query, a single where clause (e.g. `where field_a is not null`). Wheres are connected to columns as well as to queries. By connecting to queries, Amundsen can rank order the most frequent wheres by accessing the query executions.
+
+## UI/UX-level Explanation
+
+Amundsen will display the most common filters and frequent joins based off of query execution counts:
+
+Common filters:
+
+![Common Filters](../assets/037/common-filters.png)
+
+Display common joins:
+
+![Common Joins](../assets/037/common-joins.png)
+
+Queries can also be displayed for tables similar to how queries are shown for dashboards.
+
+## Reference-level Explanation (aka Technical Details)
+
+Four new models will be added to support the data assets listed above:
+
+1. Query
+    - Attributes:
+        - sql: the SQL code for the query
+        - tables: a list of Table Metadata representing each table used in the query
+        - clean_sql: an optional SQL string that the implementer has curated. For example, if an Airflow DAG injects the date into a SQL query every day, the implementer may strip that date and pass the updated SQL into `clean_sql` in order to allow Amundsen to identify that this is not a unique query.
+        - user: an optional user that executed the query (should be Amundsen user ID)
+    - Relations:
+        - Table > Query
```

Comment:
currently the dashboard model has query model already. How do we de-duplicate on table-query and dashboard-query?

also how do you plan to figure the table<->query relationship (and what is considered to have query and table relationship)? is it through sql parsing?

Reply:
> currently the dashboard model has query model already. How do we de-duplicate on table-query and dashboard-query?

good question, I'll let @sewardgw respond to that


> also how do you plan to figure the table<->query relationship (and what is considered to have query and table relationship)? is it through sql parsing?

Sort of. Stemma has full SQL parsing capabilities, but it's proprietary. Our hope is to provide the plumbing such that OSS implementors could easily sub in their own parsing capabilities. We've seen that many companies possess this capability already to varying degrees of veracity: some have simple regex based scripts that can extract basic info about the query (e.g. joins); we've seen other companies with ANTLR-based parsers with which they extract more fine-grained details.

Reply:
For SQL parsing, I wouldn't recommend `ANTLR-based` parser. Now we delete those extractors, but in the early days, we have presto Antlar parser which causes so many issues in production.

Reply:
Re: `how do you plan to figure the table<->query relationship ` We are not dictating _how_ to do this. This is definitely a complicated task but there are members of the community doing this already. We are providing the pipes to enable them to structure, access and display the information within Amundsen.

Reply:
Re: `How do we de-duplicate on table-query and dashboard-query`. Thanks for bringing this up, this is a great point. As an FYI There is also a "query" object being created as part of the ML Features PR (e.g. query used to create features.) so this question is especially relevant.

 There are some unique attributes of the dashboard query model that we do not want to lose (e.g. `dashboard_group_id`, `dashboard_id`). I think the best options are:
- One common query model that all "query" use-cases use, or
- One common Query Model that the others inherit from but add their own specific attributes (keeping the same node label / relationship labels)

My preference would be toward #2 but would love feedback before adding this additional context to the PR.

Reply:
Here is the other PR: https://github.com/amundsen-io/amundsen/pull/1059/files

Tagging @allisonsuarez here as well

Reply:
> For SQL parsing, I wouldn't recommend ANTLR-based parser. Now we delete those extractors, but in the early days, we have presto Antlar parser extractor which causes so many issues in production.

Yeah, we're not planning to build this into Amundsen -- just echo'ing what I've seen companies do in the wild

Reply:
@dorianj @feng-tao do you think the Query schema here is generic enough to be used for other resources? https://github.com/amundsen-io/amundsen/pull/1059/files#diff-4a840835d48ad91e06402d8bdcb555dcd2ed2c40457c59713ebb5c83fda64bdaR16

Reply:
I will add my 2cents in that I _do_ think the implementation from @allisonsuarez  it is generic enough (there aren't a lot of fields so the minimal implementation lends itself to be flexible). I would propose putting it in a different location though, specifically since this RFC is building out custom components of queries I would propose a query-specific file.

Reply:
it is worth noting that this is basically the same model we already use for Dashboards. I will move it to its own file so we can unblock the ML features work.